odoo/tools/safe_eval.py

```diff
@@ -10,219 +9,584 @@
+            ctx=node.ctx,
+        )
+
+    def explore_Assign(self, node: ast.Assign) -> list[ast.Assign | ast.Expr]:
+        node = cast(ast.Assign, self.generic_visit(node))
+        node.value = ast.Call(
+            func=ast.Name("__type_checker", ctx=ast.Load()),
+            args=[],
+            keywords=[
+                ast.keyword(arg="__obj", value=node.value),
+            ],
+        )
+
+        nodes: list[ast.Assign | ast.Expr] = [node]
+        for target in node.targets:
+            for n in ast.walk(cast(ast.AST, target)):
+                if isinstance(n, (ast.Tuple, ast.List)):
+                    for e in n.elts:
+                        if isinstance(e, ast.Starred):
+                            e = e.value
+
+                        nodes.append(
+                            ast.Expr(
+                                value=ast.Call(
+                                    func=ast.Name("__type_checker", ctx=ast.Load()),
+                                    args=[],
+                                    keywords=[ast.keyword(arg="__obj", value=e)],
+                                )
+                            )
+                        )
+                    break
+
+                if isinstance(n, ast.Subscript):
+                    nodes.append(
+                        ast.Expr(
+                            value=ast.Call(
+                                func=ast.Name("__type_checker", ctx=ast.Load()),
+                                args=[],
+                                keywords=[ast.keyword(arg="__obj", value=n)],
+                            )
+                        )
+                    )
+                    break
+
+                if isinstance(n, (ast.Name, ast.Attribute)):
+                    nodes.append(
+                        ast.Expr(
+                            value=ast.Call(
+                                func=ast.Name("__type_checker", ctx=ast.Load()),
+                                args=[],
+                                keywords=[ast.keyword(arg="__obj", value=n)],
+                            )
+                        )
+                    )
+                    break
+
+        return nodes
+
+    def explore_ExceptHandler(self, node: ast.ExceptHandler) -> ast.ExceptHandler:
+        node = cast(ast.ExceptHandler, self.generic_visit(node))
+        if node.name:
+            _check_attribute_name(node.name)
+
+        return node
+
+    def explore_For(self, node: ast.For) -> ast.For:
+        node = cast(ast.For, self.generic_visit(node))
+
+        node.iter = ast.Call(
+            func=ast.Name("__type_checker", ctx=ast.Load()),
+            args=[],
+            keywords=[
+                ast.keyword(arg="__obj", value=node.iter),
+            ],
+        )
+
+        if not isinstance(node.target, ast.Name):
+            for n in ast.walk(cast(ast.AST, node.target)):
+                if isinstance(n, ast.Name):
+                    node.body.append(
```

Comment:
this is useless, because you can do
```py
for a, b in [(1, 5)]:
    f(a)
```


The generated code looks like
```py
for a, b in __type_checker(__obj=[(1, 5)]):
    __call_checker(__func=f, __args=[a], __kwargs={})
    __type_checker(__obj=a)
    __type_checker(__obj=b)
```

so the checks are done after we use the objects

(you should probably insert instead of append if you want to keep that check)

Reply:
@joda-odoo this still allow me to bypass the getitem :/